runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -147,19 +155,29 @@ public Token appendCheckpoint() {
      *  @param snapshotTimestamp snapshot at which the checkpoint is taken.
      */
     @VisibleForTesting
-    public Token appendCheckpoint(Token snapshotTimestamp) {
+    public Token appendCheckpoint(Token snapshotTimestamp, Long streamTail) {
         long start = System.currentTimeMillis();
 
-        // If snapshotTimestamp (stream tail) is a negative value
+        // If stream tail is a negative value
         // we need to enforce a hole to ensure the start log address
         // is not negative (as MultiCheckpointWriter will select the min of all checkpoints)
         // Having a negative value will prevent trimming from happening.
-        if (!Address.isAddress(snapshotTimestamp.getSequence())) {
+        if (!Address.isAddress(streamTail)) {
             snapshotTimestamp = forceHole();
+            streamTail = snapshotTimestamp.getSequence();
         }
 
         try {
+            // This read attempts to ensure no future sequencer regression by materializing the log until this boundary.
             rt.getAddressSpaceView().read(snapshotTimestamp.getSequence());
+            // However, we still need to verify there has not been a sequencer regression
+            // from the time the tuple (snapshotTimestamp, streamTail) was acquired and the actual read was performed.
+            // To this end, we query the sequencer for the last observed epoch. This validation
+            // is pesimistic as we might observe the change after the read, but it is safe to retry.
+            long currentEpoch = rt.getSequencerView().query().getEpoch();
+            if (currentEpoch != snapshotTimestamp.getEpoch()) {
+                throw new WrongEpochException(currentEpoch);
+            }
         } catch (TrimmedException te) {
             // Stream Tail was trimmed, we can just ignore this trim. This means
```

Comment:
This is not a stream tail any more, it's a global tail. Can we change the comment and the log message after this?

Reply:
You know what, actually we no longer need this. This was in place because on previous design the snapshot was given by the stream's tail and this. could be trimmed, but now the snapshot is given by the global tail, for which a trimmed exception is not admitted. Removing this.

Reply:
I think this is still possible. From the time forceHole is called to the time `rt.getAddressSpaceView().read(snapshotTimestamp.getSequence());` is called, a trim can happened. Its just an unhanded exception not sure if its a serious issue.

Reply:
But if a trimmed is called in between we should actually fail right? or you suggest to include this case as part of our retry logic?